### PR TITLE
Check for opt-level="z" in `build.rs`, provide feedback about dangers

### PR DESCRIPTION
**Pull request by sethp**
With this change, trying to build with opt-level="z" produces a message that looks like this:

image

Mitigation for:

* https://github.com/esp-rs/esp-hal/issues/196
* https://github.com/llvm/llvm-project/issues/57988

### CODE REVIEW

**Comment by jessebraham** [comment]
Just adding my comment from #196:

> I think the proof of concept looks okay. I'm not entirely convinced it should be behind a feature, and I think printing a warning and continuing is probably preferable to terminating. Based on my limited testing `opt-level = "z"` _does_ work for a number of examples, so I think this case should be treated as the exception and not the norm. Happy to hear any arguments for alternatives, of course.

**Comment by bjoernQ** [comment]
My two cents: I would be fine with aborting the build by default. Especially for the examples we don't configure `opt-level="z"` by default. Also having a way to opt-out of the check is fine I guess

I agree that warnings are easy to miss like when there are a lot of other warnings later during the build

File: esp32c3-hal/build.rs [hunk 1]
```diff
@@ -71,3 +75,41 @@ fn add_defaults() {
 
     println!("cargo:rustc-link-search={}", out.display());
 }
+
+const OPT_LEVEL_Z_MSG: &str = r#"opt-level=z will produce broken 128-bit shifts (i.e. `1u128 << i`). The hal's interrupt handling relies on that operation, causing an 'attempt to subtract with overflow' panic if an enabled interrupt is triggered while using that opt-level.
+
+Please use `opt-level="s"` in lieu of "z", or alternatively enable `features = ["allow-opt-level-z"]` to supress this error. The latter option is only recommended if you:
```
Comment (gustavonihei) [inline esp32c3-hal/build.rs:81]:
```suggestion
Please use `opt-level="s"` in lieu of "z", or alternatively enable `features = ["allow-opt-level-z"]` to suppress this error. The latter option is only recommended if you:
```

**Review by bjoernQ — APPROVED** [review]
LGTM
Thanks for the PR - sorry it took a bit longer to get in

**Comment by MabezDev** [comment]
We may be able to revert this soon, looks like it might have been fixed in compiler builtins: https://github.com/rust-lang/compiler-builtins/pull/521?

**Comment by sethp** [comment]
Oh wow, excellent news! Good timing, too: I was just looking for `opt-level=z` the other day.

Do you have a sense for what the release flow looks like for `compiler-builtins`? Is it a "any nightly 2023-03-24 or later is safe to use with `opt-level=z`" kind of situation?

**Comment by MabezDev** [comment]
I _think_ this has now landed. I'm not sure how to get the version of compiler_builtins used in core, but at least in std they are using 0.191, and the fix should be included from 0.189 onwards. I think we can revert this now.

**Pull request merged**